<?xml version="1.0" encoding="UTF-8"?>
<launch>
<!-- Reference: -->
<!-- https://github.com/Toronto-Robotics-club/ros_control_and_gazebo_simple_position_controller.git -->

    <!-- add the Gazebo world so that we launch it all together-->
    <arg name="model" default="$(find xy_platform_control)/urdf/xy_plotter.urdf"/>
    <arg name="world" default="empty"/> 
    <arg name="paused" default="false"/>
    <arg name="use_sim_time" default="true"/>
    <arg name="gui" default="true"/>
    <arg name="headless" default="false"/>
    <arg name="debug" default="false"/>
    <arg name="verbose" value="true"/>

    <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <arg name="world_name" value="$(find xy_platform_control)/worlds/simple.world"/>
    <arg name="paused" value="$(arg paused)"/>
    <arg name="use_sim_time" value="$(arg use_sim_time)"/>
    <arg name="gui" value="$(arg gui)"/>
    <arg name="headless" value="$(arg headless)"/>
    <arg name="debug" value="$(arg debug)"/>
    <!--by activating "verbose" mode, Gazebo prints more to terminal on launch.-->
    <arg name="verbose" value="$(arg verbose)"/>
    </include>

    <!--load urdf into parameter server-->
    <param name="robot_description" textfile="$(find xy_platform_control)/urdf/xy_plotter.urdf" />
    
    <!-- spawn xy plotter -->
    <node name="urdf_spawner" 
          pkg="gazebo_ros" 
          type="spawn_model" 
          respawn="false" 
          output="screen" 
          args="-urdf -model xy_plotter -param robot_description"/>

    <!-- load the controllers -->
    <rosparam file="$(find xy_platform_control)/config/xyplatform_sim_config.yaml" command="load"/>

    <node name="controller_spawner" 
          pkg ="controller_manager" 
          type="spawner" 
          ns="/xy_plotter" 
          args="base_to_slider_joint_position_controller slider_to_second_joint_position_controller joint_state_controller"/>
    
	<!-- converts joint states to TF transforms -->
    <node name="robot_state_publisher" 
          pkg="robot_state_publisher" 
          type="robot_state_publisher" 
          respawn="false" 
          output="screen">
        <remap from="joint_states" to="/xy_plotter/joint_states" />
    </node>

    <!-- node to control the xy_platform in gazebo -->
    <node name="xy_platform_controller" pkg="xy_platform_control" type="xy_platform_controller.py" output="screen"/>

    <!-- Launch RViz -->
    <node name="rviz" pkg="rviz" type="rviz" args="-d $(find xy_platform_control)/rviz/xy_platform.rviz" output="screen" />

</launch>

